AddQtAndroidApk.cmake

```diff
@@ -97,84 +98,162 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
+        set(QT_ANDROID_APP_PACKAGE_SOURCE_IN ${ARG_PACKAGE_SOURCES})
 
+        set(QT_ANDROID_APP_PACKAGE_SOURCE_ROOT "${CMAKE_CURRENT_BINARY_DIR}/package")
+        set(QT_ANDROID_MANIFEST_SOURCE_IN "${CMAKE_CURRENT_BINARY_DIR}/packagein")
+
+        configure_file(${QT_ANDROID_MANIFEST_IN_REAL_PATH} ${QT_ANDROID_MANIFEST_SOURCE_IN}/AndroidManifest.xml @ONLY)
+        endif()
+   else()
         # create a subdirectory for the extra package sources
         set(QT_ANDROID_APP_PACKAGE_SOURCE_ROOT "${CMAKE_CURRENT_BINARY_DIR}/package")
 
         # generate a manifest from the template
-        configure_file(${QT_ANDROID_SOURCE_DIR}/AndroidManifest.xml.in ${QT_ANDROID_APP_PACKAGE_SOURCE_ROOT}/AndroidManifest.xml @ONLY)
+        configure_file(${QT_ANDROID_MANIFEST_IN_REAL_PATH} ${QT_ANDROID_APP_PACKAGE_SOURCE_ROOT}/AndroidManifest.xml @ONLY)
+    endif()
+
+    # newer NDK toolchains don't define ANDROID_STL_PREFIX anymore,
+    # so this is a fallback to the only supported value in recent versions
+    if(NOT ANDROID_STL_PREFIX)
+        if(ANDROID_STL MATCHES "^c\\+\\+_")
+            set(ANDROID_STL_PREFIX llvm-libc++)
+        endif()
+    endif()
+
+    if(NOT ANDROID_STL_PREFIX)
+        message(WARNING "Failed to determine ANDROID_STL_PREFIX value for ANDROID_STL=${ANDROID_STL}.")
     endif()
 
     # define the STL shared library path
+    # up until NDK r18, ANDROID_STL_SHARED_LIBRARIES is populated by the NDK's toolchain file
+    # since NDK r19, the only option for a shared STL library is libc++_shared
     if(ANDROID_STL_SHARED_LIBRARIES)
         list(GET ANDROID_STL_SHARED_LIBRARIES 0 STL_LIBRARY_NAME) # we can only give one to androiddeployqt
         if(ANDROID_STL_PATH)
-            set(QT_ANDROID_STL_PATH "${ANDROID_STL_PATH}/libs/${ANDROID_ABI}/lib${STL_LIBRARY_NAME}.so")
+            set(QT_ANDROID_STL_PATH "${ANDROID_STL_PATH}/libs/${ANDROID_ABI}/lib${ANDROID_STL}.so")
         else()
-            set(QT_ANDROID_STL_PATH "${ANDROID_NDK}/sources/cxx-stl/${ANDROID_STL_PREFIX}/libs/${ANDROID_ABI}/lib${STL_LIBRARY_NAME}.so")
+            set(QT_ANDROID_STL_PATH "${QT_ANDROID_NDK_ROOT}/sources/cxx-stl/${ANDROID_STL_PREFIX}/libs/${ANDROID_ABI}/lib${ANDROID_STL}.so")
         endif()
+    elseif(ANDROID_STL STREQUAL c++_shared)
+        set(QT_ANDROID_STL_PATH "${QT_ANDROID_NDK_ROOT}/sources/cxx-stl/${ANDROID_STL_PREFIX}/libs/${ANDROID_ABI}/libc++_shared.so")
     else()
         set(QT_ANDROID_STL_PATH)
+        if(ANDROID_STL_STATIC_LIBRARIES)
+            message(WARNING "ANDROID_STL_SHARED_LIBRARIES isn't defined, you might need to define ANDROID_STL (${ANDROID_STL}) to a shared stl library and not a static library")
+        else(ANDROID_STL_STATIC_LIBRARIES)
+            message(WARNING "ANDROID_STL_SHARED_LIBRARIES isn't defined, you might need to define ANDROID_STL (${ANDROID_STL}) to a shared stl library (for example c++_shared or gnustl_shared)")
+        endif(ANDROID_STL_STATIC_LIBRARIES)
     endif()
```

Comment:
This STL part is not clear to me. With default settings (and NDK version 18) I get a warning:

>CMake Warning at dashboard-app/3rdparty/qt-android-cmake/AddQtAndroidApk.cmake:188 (message):
  ANDROID_STL_SHARED_LIBRARIES isn't defined, you might need to define
  ANDROID_STL (c++_static) to a shared stl library and not a static library
Call Stack (most recent call first):
  dashboard-app/CMakeLists.txt:52 (add_qt_android_apk)

Why is `ANDROID_STL_SHARED_LIBRARIES` undefined? Why do I have to define `ANDROID_STL` to a shared STL, and what happens if I don't?

Can't this thing get an implementation that "just works" by default? Or maybe I just need to upgrade my NDK? 😃

Reply:
I don't really remember about this, but i remember it was a mess when i updated to NDK 19.
For example my cmake command look like this:
```
cmake -DCMAKE_PREFIX_PATH=C:/Qt/5.12.1/android_armv7 \
-G "Unix Makefiles" \
-DANDROID_ABI=armeabi-v7a \
-DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake \
-DANDROID_NATIVE_API_LEVEL=27 \
-DANDROID_STL=c++_shared \
-DANDROID_TOOLCHAIN=clang
```
So i as i remember from NDK 19 you need to use `c++_shared`. And if i remember correctly with NDK 18 it was `gnustl_shared`. But i could be totally wrong here.

Reply:
> So i as i remember from NDK 19 you need to use c++_shared. And if i remember correctly with NDK 18 it was gnustl_shared

What I don't understand is, why aren't these the defaults? Is it Qt that forces to use a shared STL? If so, would it be a good idea to set a proper default in the CMake script, according to what you say?

Reply:
So i tried to see if gnustl_shared was still working and it seems not:
```
CMake Error at C:/Users/olivi/AppData/Local/Android/Sdk/android-ndk-r20b/build/cmake/android.toolchain.cmake:249 (message):
  gnustl_shared is no longer supported.  Please switch to either c++_shared
  or c++_static.  See
  https://developer.android.com/ndk/guides/cpp-support.html for more
  information.
```

Reply:
I tried `c++_static` and i get:
```
Invalid STD C++ library name.
```
When it tries to build to build the APK

Reply:
Ok so with NDK20 on my machine with the toolchain i have ANDROID_STL by default set to c++_static. i fixed the script to generate `"stdcpp-path": "C:/Users/olivi/AppData/Local/Android/Sdk/android-ndk-r20b/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a/libc++_static.a",` inside qtdeploy.json.
But it seem that qtandroiddeploy doesn't care about that (at least the .a suffix). I end up in my apk with a file libc++_static.so that is a copy of the correct libc++_static.a.
When i try so launch the app i get the error
```
com.oliv.qaterial E/AndroidRuntime: FATAL EXCEPTION: qtMainLoopThread
    Process: com.oliv.qaterial, PID: 27050
    java.lang.UnsatisfiedLinkError: dlopen failed: "/data/app/com.oliv.qaterial-0EE82Hk--0TzcAuFs5v6Yw==/lib/arm/libc++_static.so" has bad ELF magic
        at java.lang.Runtime.load0(Runtime.java:928)
        at java.lang.System.load(System.java:1633)
        at org.qtproject.qt5.android.QtNative$2.run(QtNative.java:199)
        at org.qtproject.qt5.android.QtThread$2.run(QtThread.java:87)
        at org.qtproject.qt5.android.QtThread$1.run(QtThread.java:61)
        at java.lang.Thread.run(Thread.java:764)
```
It seem that qt for android is compiled by default with c++_shared stl. Problem is that ANDROID_STL is default to c++_static in the NDK toolchain file.
I don't see any other fix than continuing specifying ANDROID_STL=c++_shared when doing the first cmake command.

I will add a warning that say that ANDROID_STL must be set to c++_shared for now.

Reply:
Ok so i fixed it in 1adbdc5575dc6780285de6833da0a071e415e272
Qt is build with c++_shared stl and is expecting it. Like qtandroiddeploy. So even if ANDROID_STL isn't set to c++_shared, libc++_shared.so will be deployed inside the apk with a warning for the dev. 
This will work just fine if ANDROID_STL is the default value c++_static. In that scenario, YourApp.so will be build using c++_static and qt*.so libraries are going to find the required shared stl.

Reply:
Thanks.

> `"You should consider setting ANDROID_STL to c++_shared (like qt)."`

It's Qt, not qt 😆

> "Note that gnustl_shared is no longer supported"

Idn't it redundant with the warning from the toolchain file?

Reply:
You're right.
Fixed in 7ab6d6cfc73764fd6e21d663116d3b8cfa3a7530